ui/js/graphs.js

```diff
@@ -801,6 +814,71 @@ perf.controller('TestChooserCtrl', function($scope, $modalInstance, $http,
         });
     };
 
+    var resolvePromise = $q.defer();
+    var addRelatedPlatforms = function(relatedSeries) {
+        PhSeries.getAllSeries(relatedSeries.projectName,
+            $scope.timeRange, optionCollectionMap).then(
+            function(seriesData) {
+                var platformList = seriesData.platformList;
+                platformList.forEach(function(platform) {
+                    // query all the tests with specific platform, then find
+                    // out the test which has same with the existed test
+                    var testList = _.sortBy(_.filter(seriesData.seriesList,
+                        { platform: platform }), 'name');
+                    var temp = _.findWhere(testList, {"name": relatedSeries.name});
+                    // if found something different from the series we already have,
+                    // then we push it into the testsToAdd list.
+                    if (temp !== undefined && temp.signature !== relatedSeries.signature) {
+                        $scope.testsToAdd.push(_.clone(temp));
+                    }
+                });
+            }
+        ).then(function() {
+                // resolve the testsToAdd's length after every thing was done
+                // so we don't need timeout here
+                resolvePromise.resolve($scope.testsToAdd.length);
```

Comment:
Due to the the remote request had timeout function inside, so we only care about whether we found related data or not. Therefore, we could just resolve the $scope.testsToAdd.length into the promise. If the length is more than zero(mean we had found something related and add them to the list) we should display the testchooser, or show the alert window because nothing been added.
Do you agree with it, Will? Please tell me if you think it's second preference choice, thank you :)